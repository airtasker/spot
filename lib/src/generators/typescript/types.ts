import assertNever from "assert-never";
import * as ts from "typescript";
import {
  Api,
  ArrayType,
  normalizedObjectType,
  ObjectType,
  OptionalType,
  Type,
  TypeReference,
  Types,
  UnionType
} from "../../models";
import { outputTypeScriptSource } from "./ts-writer";
import compact = require("lodash/compact");

export function generateTypesSource(api: Api): string {
  return outputTypeScriptSource(
    Object.entries(api.types).map(([typeName, type]) =>
      ts.createTypeAliasDeclaration(
        /*decorators*/ undefined,
        /*modifiers*/ [ts.createModifier(ts.SyntaxKind.ExportKeyword)],
        typeName,
        /*typeParameters*/ undefined,
        typeNode(api.types, type)
      )
    )
  );
}

export function promiseTypeNode(types: Types, type: Type): ts.TypeNode {
  return ts.createTypeReferenceNode("Promise", [typeNode(types, type)]);
}

export function typeNode(types: Types, type: Type): ts.TypeNode {
  switch (type.kind) {
    case "void":
      return VOID_TYPE_NODE;
    case "null":
      return NULL_TYPE_NODE;
    case "boolean":
      return BOOLEAN_TYPE_NODE;
    case "boolean-constant":
      return type.value ? TRUE_TYPE_NODE : FALSE_TYPE_NODE;
    case "date":
    case "date-time":
    case "string":
      return STRING_TYPE_NODE;
    case "string-constant":
      return stringConstantTypeNode(type.value);
    case "number":
    case "float":
    case "double":
    case "int32":
    case "int64":
      return NUMBER_TYPE_NODE;
    case "integer-constant":
      return integerConstantTypeNode(type.value);
    case "object":
      return objectTypeNode(types, type);
    case "array":
      return arrayTypeNode(types, type);
    case "optional":
      return optionalTypeNode(types, type);
    case "union":
      return unionTypeNode(types, type);
    case "type-reference":
      return typeReferenceTypeNode(type);
    default:
      throw assertNever(type);
  }
}

const VOID_TYPE_NODE = ts.createKeywordTypeNode(ts.SyntaxKind.VoidKeyword);

const NULL_TYPE_NODE = ts.createKeywordTypeNode(ts.SyntaxKind.NullKeyword);

const BOOLEAN_TYPE_NODE = ts.createKeywordTypeNode(
  ts.SyntaxKind.BooleanKeyword
);

const TRUE_TYPE_NODE = ts.createLiteralTypeNode(ts.createLiteral(true));
const FALSE_TYPE_NODE = ts.createLiteralTypeNode(ts.createLiteral(false));

const STRING_TYPE_NODE = ts.createKeywordTypeNode(ts.SyntaxKind.StringKeyword);

const NUMBER_TYPE_NODE = ts.createKeywordTypeNode(ts.SyntaxKind.NumberKeyword);

function stringConstantTypeNode(value: string): ts.TypeNode {
  return ts.createLiteralTypeNode(ts.createLiteral(value));
}

function integerConstantTypeNode(value: number): ts.TypeNode {
  return ts.createLiteralTypeNode(ts.createLiteral(value));
}

function objectTypeNode(types: Types, type: ObjectType): ts.TypeNode {
  return ts.createTypeLiteralNode(
    Object.entries(normalizedObjectType(types, type)).map(
      ([propertyName, propertyType]) => {
        if (propertyType.kind === "void") {
          return ts.createPropertySignature(
            /*modifiers*/ undefined,
            propertyName,
            ts.createToken(ts.SyntaxKind.QuestionToken),
            VOID_TYPE_NODE,
            /*initializer*/ undefined
          );
        } else if (propertyType.kind === "optional") {
          return ts.createPropertySignature(
            /*modifiers*/ undefined,
            propertyName,
            ts.createToken(ts.SyntaxKind.QuestionToken),
            typeNode(types, propertyType.optional),
            /*initializer*/ undefined
          );
        } else {
          return ts.createPropertySignature(
            /*modifiers*/ undefined,
            propertyName,
            /*questionToken*/ undefined,
            typeNode(types, propertyType),
            /*initializer*/ undefined
          );
        }
      }
    )
  );
}

function arrayTypeNode(types: Types, type: ArrayType): ts.TypeNode {
  return ts.createArrayTypeNode(typeNode(types, type.elements));
}

function optionalTypeNode(types: Types, type: OptionalType): ts.TypeNode {
  return ts.createUnionTypeNode([
    typeNode(types, type.optional),
    VOID_TYPE_NODE
  ]);
}

function unionTypeNode(types: Types, type: UnionType): ts.TypeNode {
  return ts.createUnionTypeNode(type.types.map(t => typeNode(types, t)));
}

function typeReferenceTypeNode(type: TypeReference): ts.TypeNode {
  return ts.createTypeReferenceNode(type.typeName, /*typeArguments*/ undefined);
}
