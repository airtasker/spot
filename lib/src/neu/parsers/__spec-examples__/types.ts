import {
  Date,
  DateTime,
  Double,
  Float,
  Int32,
  Int64,
  Integer,
  Number,
  String
} from "@airtasker/spot";

interface TypeInterface {
  null: null;
  boolean: boolean;
  true: true;
  false: false;
  string: string;
  String: String;
  literalString: "literalString";
  number: number;
  Number: Number;
  Float: Float;
  Double: Double;
  literalFloat: 1.02;
  Integer: Integer;
  Int32: Int32;
  Int64: Int64;
  literalInteger: 2;
  Date: Date;
  DateTime: DateTime;
  literalObject: {
    propertyA: string;
    propertyB?: boolean;
  };
  array: boolean[];
  union: boolean | Date | null;
  alias: Alias;
  interface: Interface;
  interfaceExtends: InterfaceExtends;
  indexedAccess: IndexedAccess["root"];
  indexedAccessNested: IndexedAccess["child"]["nested"]["secondNest"];
  indexedIndexedAccess: IndexedAccess["indexed"]["root"];
  indexedAccessInline: { root: boolean }["root"];
  enum: Enum;
  enumConstant: Enum.A;
  map: Map<string, boolean>;
  indexSignature: { [index: string]: boolean };
  interfaceWithIndexSignature: InterfaceWithIndexSignature;
}

type Alias = string;

interface Interface {
  interfaceProperty: boolean;
}

interface InterfaceExtends extends Interface {
  interfaceExtendsProperty: boolean;
}

interface InterfaceWithIndexSignature {
  [index: string]: boolean;
}

interface IndexedAccess {
  root: boolean;
  child: {
    nested: {
      secondNest: boolean;
    };
  };
  indexed: IndexedIndexedAccess;
}

interface IndexedIndexedAccess {
  root: boolean;
}

enum Enum {
  A,
  B,
  C
}
