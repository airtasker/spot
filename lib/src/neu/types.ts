enum TypeKind {
  NULL = "null",
  BOOLEAN = "boolean",
  BOOLEAN_LITERAL = "boolean-literal",
  STRING = "string",
  STRING_LITERAL = "string-literal",
  FLOAT = "float",
  DOUBLE = "double",
  FLOAT_LITERAL = "float-literal",
  INT32 = "int32",
  INT64 = "int64",
  INT_LITERAL = "integer-literal",
  DATE = "date",
  DATE_TIME = "date-time",
  OBJECT = "object",
  ARRAY = "array",
  UNION = "union",
  REFERENCE = "reference"
}

export type Type =
  | NullType
  | BooleanType
  | BooleanLiteralType
  | StringType
  | StringLiteralType
  | FloatType
  | DoubleType
  | FloatLiteralType
  | Int32Type
  | Int64Type
  | IntLiteralType
  | DateType
  | DateTimeType
  | ObjectType
  | ArrayType
  | UnionType
  | ReferenceType;

export interface NullType {
  kind: TypeKind.NULL;
}

export interface BooleanType {
  kind: TypeKind.BOOLEAN;
}

export interface BooleanLiteralType {
  kind: TypeKind.BOOLEAN_LITERAL;
  value: boolean;
}

export interface StringType {
  kind: TypeKind.STRING;
}

export interface StringLiteralType {
  kind: TypeKind.STRING_LITERAL;
  value: string;
}

export interface FloatType {
  kind: TypeKind.FLOAT;
}

export interface DoubleType {
  kind: TypeKind.DOUBLE;
}

export interface FloatLiteralType {
  kind: TypeKind.FLOAT_LITERAL;
  value: number;
}

export interface Int32Type {
  kind: TypeKind.INT32;
}

export interface Int64Type {
  kind: TypeKind.INT64;
}

export interface IntLiteralType {
  kind: TypeKind.INT_LITERAL;
  value: number;
}

export interface DateType {
  kind: TypeKind.DATE;
}

export interface DateTimeType {
  kind: TypeKind.DATE_TIME;
}

export interface ObjectType {
  kind: TypeKind.OBJECT;
  properties: Array<{
    name: string;
    description?: string;
    optional: boolean;
    type: Type;
  }>;
}

export interface ArrayType {
  kind: TypeKind.ARRAY;
  elementType: Type;
}

export interface UnionType {
  kind: TypeKind.UNION;
  types: Type[];
}

export interface ReferenceType {
  kind: TypeKind.REFERENCE;
  name: string;
}

export function nullType(): NullType {
  return {
    kind: TypeKind.NULL
  };
}

export function booleanType(): BooleanType {
  return {
    kind: TypeKind.BOOLEAN
  };
}

export function booleanLiteralType(value: boolean): BooleanLiteralType {
  return {
    kind: TypeKind.BOOLEAN_LITERAL,
    value
  };
}

export function stringType(): StringType {
  return {
    kind: TypeKind.STRING
  };
}

export function stringLiteralType(value: string): StringLiteralType {
  return {
    kind: TypeKind.STRING_LITERAL,
    value
  };
}

export function floatType(): FloatType {
  return {
    kind: TypeKind.FLOAT
  };
}

export function doubleType(): DoubleType {
  return {
    kind: TypeKind.DOUBLE
  };
}

export function floatLiteralType(value: number): FloatLiteralType {
  return {
    kind: TypeKind.FLOAT_LITERAL,
    value
  };
}

export function int32Type(): Int32Type {
  return {
    kind: TypeKind.INT32
  };
}

export function int64Type(): Int64Type {
  return {
    kind: TypeKind.INT64
  };
}

export function intLiteralType(value: number): IntLiteralType {
  return {
    kind: TypeKind.INT_LITERAL,
    value
  };
}

export function dateType(): DateType {
  return {
    kind: TypeKind.DATE
  };
}

export function dateTimeType(): DateTimeType {
  return {
    kind: TypeKind.DATE_TIME
  };
}

export function objectType(
  properties: Array<{
    name: string;
    description?: string;
    optional: boolean;
    type: Type;
  }>
): ObjectType {
  return {
    kind: TypeKind.OBJECT,
    properties
  };
}

export function arrayType(elementType: Type): ArrayType {
  return {
    kind: TypeKind.ARRAY,
    elementType
  };
}

export function unionType(unionTypes: Type[]): UnionType {
  return {
    kind: TypeKind.UNION,
    types: unionTypes
  };
}

export function referenceType(name: string): ReferenceType {
  return {
    kind: TypeKind.REFERENCE,
    name
  };
}

/**
 * Loci table is a lookup table for types.
 */
export class TypeTable {
  private types: Map<string, Type>;

  constructor(types: Map<string, Type> = new Map<string, Type>()) {
    this.types = types;
  }

  /**
   * Return an object representation of the type table.
   */
  toArray(): Array<{ name: string; type: Type }> {
    const arr = new Array<{ name: string; type: Type }>();
    this.types.forEach((type, key, _) => {
      arr.push({ name: key, type });
    });
    return arr.sort((a, b) => (b.name > a.name ? -1 : 1));
  }

  /**
   * Add a type to the type table. If the type key is already present, `add` will throw an error.
   *
   * @param key lookup key
   * @param locus target locus
   */
  add(key: string, type: Type): void {
    if (this.types.has(key)) {
      throw new Error(`Key already present in type table: ${key}`);
    }
    this.types.set(key, type);
  }

  /**
   * Check if a type exists in the table.
   *
   * @param key lookup key
   */
  exists(key: string): boolean {
    return this.types.has(key);
  }
}
