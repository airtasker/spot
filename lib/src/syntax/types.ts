/** A float */
export type Float = number;

/** An integer */
export type Integer = number;

/** A string */
export type String = string;

/**
 * A `full-date` as defined by https://tools.ietf.org/html/rfc3339#section-5.6
 *
 * @example
 * "2018-08-24"
 */
export type Date = string;

/**
 * A `date-time` as defined by https://tools.ietf.org/html/rfc3339#section-5.6
 *
 * @example
 * "2018-08-24T21:18:36Z"
 */
export type DateTime = string;
