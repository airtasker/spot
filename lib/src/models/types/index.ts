import {
  CustomPrimitiveType,
  DateTimeType,
  DateType,
  IntegerType
} from "./custom-primitive-types";
import { TypeKind } from "./kinds";
import { ArrayType, ObjectType } from "./object-types";
import {
  NumberLiteral,
  PrimitiveLiteral,
  StringLiteral
} from "./primitive-literals";
import { NumberType, PrimitiveType, StringType } from "./primitive-types";
import { ReferenceType, UnionType } from "./special-types";

export * from "./custom-primitive-types";
export * from "./kinds";
export * from "./object-types";
export * from "./primitive-literals";
export * from "./primitive-types";
export * from "./special-types";

export type DataType =
  | PrimitiveType
  | CustomPrimitiveType
  | PrimitiveLiteral
  | ObjectType
  | ArrayType
  | ReferenceType
  | UnionType;

export type NumberLikeType =
  | NumberType
  | NumberLiteral
  | IntegerType
  | ReferenceType;

export type StringLikeType =
  | StringType
  | StringLiteral
  | DateType
  | DateTimeType
  | ReferenceType;

export function isObjectType(type: DataType): type is ObjectType {
  return type.kind === TypeKind.OBJECT;
}

export function isArrayType(type: DataType): type is ArrayType {
  return type.kind === TypeKind.ARRAY;
}

export function isUnionType(type: DataType): type is UnionType {
  return type.kind === TypeKind.UNION;
}

export function isReferenceType(type: DataType): type is ReferenceType {
  return type.kind === TypeKind.TYPE_REFERENCE;
}
