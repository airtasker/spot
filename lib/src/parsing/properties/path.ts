import ts from "typescript";
import { DynamicPathComponent, PathComponent, VOID } from "../../models";
import { extractSingleDecorator } from "../decorators";
import {
  isObjectLiteral,
  isStringLiteral,
  ObjectLiteral
} from "../literal-parser";
import { panic } from "../panic";
import { extractParameterType } from "./parameter-type";

/**
 * Returns the path of an endpoint.
 *
 * For example:
 * ```
 * @endpoint({
 *   method: "GET",
 *   path: "/users/:userId"
 * })
 * getUser(@pathParam() userId: string): User {
 *   ...
 * }
 * ```
 *
 * will return a list of path components with:
 * - a static component ("/users/")
 * - a dynamic component named userId of type string
 */
export function extractPath(
  sourceFile: ts.SourceFile,
  methodDeclaration: ts.MethodDeclaration,
  endpointDescription: ObjectLiteral
): PathComponent[] {
  const pathLiteral = endpointDescription.properties["path"];
  if (!isStringLiteral(pathLiteral)) {
    throw panic(
      `Invalid path in endpoint description: ${methodDeclaration.getText(
        sourceFile
      )}`
    );
  }
  const path = pathLiteral.text;
  const pathComponents: PathComponent[] = [];
  const dynamicPathComponents: { [name: string]: DynamicPathComponent } = {};
  if (path.length > 0) {
    let componentStartPosition = 0;
    do {
      if (path.charAt(componentStartPosition) === ":") {
        // The parameter name extends until a character that isn't a valid name.
        const nextNonNamePositionRelative = path
          .substr(componentStartPosition + 1)
          .search(/[^a-z0-9_]/gi);
        const dynamicPathComponent: DynamicPathComponent = {
          kind: "dynamic",
          name: path.substr(
            componentStartPosition + 1,
            nextNonNamePositionRelative === -1
              ? undefined
              : nextNonNamePositionRelative
          ),
          type: VOID,
          description: ""
        };
        pathComponents.push(dynamicPathComponent);
        dynamicPathComponents[dynamicPathComponent.name] = dynamicPathComponent;
        componentStartPosition =
          nextNonNamePositionRelative === -1
            ? -1
            : componentStartPosition + 1 + nextNonNamePositionRelative;
      } else {
        // The static component extends until the next parameter, which starts with ":".
        const nextColumnPosition = path.indexOf(":", componentStartPosition);
        pathComponents.push({
          kind: "static",
          content: path.substring(
            componentStartPosition,
            nextColumnPosition === -1 ? undefined : nextColumnPosition
          )
        });
        componentStartPosition = nextColumnPosition;
      }
    } while (componentStartPosition !== -1);
  }
  for (const parameter of methodDeclaration.parameters) {
    processPathParameter(
      sourceFile,
      methodDeclaration,
      parameter,
      dynamicPathComponents
    );
  }
  return pathComponents;
}

function processPathParameter(
  sourceFile: ts.SourceFile,
  methodDeclaration: ts.MethodDeclaration,
  parameter: ts.ParameterDeclaration,
  dynamicPathComponents: { [name: string]: DynamicPathComponent }
) {
  const pathParamDecorator = extractSingleDecorator(
    sourceFile,
    parameter,
    "pathParam"
  );
  if (!pathParamDecorator) {
    return;
  }
  const name = parameter.name.getText(sourceFile);
  if (dynamicPathComponents[name]) {
    dynamicPathComponents[name].type = extractParameterType(
      sourceFile,
      parameter
    );
  } else {
    throw panic(
      `Found a path parameter that isn't present in path. Expected one of [${Object.keys(
        dynamicPathComponents
      ).join(", ")}], got this instead: ${name}`
    );
  }
  if (pathParamDecorator.arguments.length > 1) {
    throw panic(
      `Expected exactly one or no arguments for @pathParam(), got ${
        pathParamDecorator.arguments.length
      }`
    );
  }
  if (pathParamDecorator.arguments.length === 1) {
    const pathParamDescription = pathParamDecorator.arguments[0];
    if (!isObjectLiteral(pathParamDescription)) {
      throw panic(
        `@pathParam() expects an object literal, got this instead: ${methodDeclaration.getText(
          sourceFile
        )}`
      );
    }
    const descriptionProperty = pathParamDescription.properties["description"];
    if (!descriptionProperty || !isStringLiteral(descriptionProperty)) {
      throw panic(
        `@pathParam() expects a string description, got this instead: ${methodDeclaration.getText(
          sourceFile
        )}`
      );
    }
    dynamicPathComponents[name].description = descriptionProperty.text;
  }
}
