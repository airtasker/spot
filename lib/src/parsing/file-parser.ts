import * as fs from "fs-extra";
import * as path from "path";
import * as ts from "typescript";
import { Api } from "../models";
import { validate } from "../validator";
import { parseApiClass } from "./api-class-parser";
import { extractSingleDecorator } from "./decorators";
import { panic } from "./panic";
import { extractObjectType, extractType } from "./type-parser";

/**
 * Parses a TypeScript source file, as well as any other TypeScript files it imports recursively.
 *
 * This should be smart enough not to get stuck into an infinite loop for circular dependencies.
 */
export async function parsePath(sourcePath: string): Promise<Api> {
  const api: Api = {
    endpoints: {},
    types: {}
  };
  await parseFileRecursively(api, new Set(), sourcePath);
  const errors = validate(api);
  if (errors.length > 0) {
    throw panic(errors.join("\n"));
  }
  return api;
}

async function parseFileRecursively(
  api: Api,
  visitedPaths: Set<string>,
  sourcePath: string
): Promise<void> {
  if (!(await fs.existsSync(sourcePath))) {
    if (await fs.existsSync(sourcePath + ".ts")) {
      sourcePath += ".ts";
    } else {
      throw panic(`No source file found at ${sourcePath}`);
    }
  }
  if (
    visitedPaths.has(sourcePath) ||
    path.resolve(sourcePath).startsWith(__dirname)
  ) {
    return;
  } else {
    visitedPaths.add(sourcePath);
  }
  const fileContent = await fs.readFile(sourcePath, "utf8");
  const sourceFile = ts.createSourceFile(
    path.basename(sourcePath),
    fileContent,
    ts.ScriptTarget.Latest
  );
  for (const statement of sourceFile.statements) {
    if (ts.isImportDeclaration(statement)) {
      if (!ts.isStringLiteral(statement.moduleSpecifier)) {
        throw panic(
          `Unsupported import statement: ${statement.moduleSpecifier.getText(
            sourceFile
          )}`
        );
      }
      const importPath = statement.moduleSpecifier.text;
      if (!importPath.startsWith(".")) {
        // This is not a relative import, we'll ignore it.
        continue;
      }
      await parseFileRecursively(
        api,
        visitedPaths,
        path.join(sourcePath, "..", importPath)
      );
    } else if (ts.isClassDeclaration(statement)) {
      const apiDecorator = extractSingleDecorator(sourceFile, statement, "api");
      if (apiDecorator) {
        parseApiClass(sourceFile, statement, api);
      }
    } else if (ts.isTypeAliasDeclaration(statement)) {
      const name = statement.name.getText(sourceFile);
      api.types[name] = extractType(sourceFile, statement.type);
    } else if (ts.isInterfaceDeclaration(statement)) {
      const name = statement.name.getText(sourceFile);
      api.types[name] = extractObjectType(sourceFile, statement);
    }
  }
}
