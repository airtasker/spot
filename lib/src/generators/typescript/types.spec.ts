import * as fs from "fs-extra";
import * as path from "path";
import {
  arrayType,
  BOOLEAN,
  booleanConstant,
  integerConstant,
  NULL,
  NUMBER,
  objectType,
  optionalType,
  STRING,
  stringConstant,
  typeReference,
  unionType,
  VOID
} from "../../models";
import { parsePath } from "../../parsing/file-parser";
import { generateTypesSource } from "./types";

const EXAMPLES_DIR = path.join(
  __dirname,
  "..",
  "..",
  "..",
  "..",
  "examples",
  "src"
);

describe("TypeScript types generator", () => {
  describe("produces valid code", () => {
    for (const testCaseName of fs.readdirSync(EXAMPLES_DIR)) {
      if (!fs.lstatSync(path.join(EXAMPLES_DIR, testCaseName)).isDirectory()) {
        continue;
      }
      test(testCaseName, async () => {
        const api = await parsePath(
          path.join(EXAMPLES_DIR, testCaseName, `${testCaseName}-api.ts`)
        );
        const source = generateTypesSource(api);
        expect(source).toMatchSnapshot();
      });
    }
  });

  it("does not generate types for endpoints", () => {
    expect(
      generateTypesSource({
        endpoints: {
          example: {
            method: "POST",
            path: [],
            description: "",
            headers: {},
            queryParams: [],
            requestType: STRING,
            responseType: STRING,
            genericErrorType: STRING,
            specificErrorTypes: {
              forbidden: {
                statusCode: 403,
                type: STRING
              },
              notFound: {
                statusCode: 404,
                type: STRING
              }
            }
          }
        },
        types: {},
        description: {
          name: "name",
          description: "description"
        }
      })
    ).toMatchInlineSnapshot(`""`);
  });

  describe("generates type validator", () => {
    test("void", () => {
      expect(
        generateTypesSource({
          endpoints: {},
          types: {
            Example: VOID
          },
          description: {
            name: "name",
            description: "description"
          }
        })
      ).toMatchInlineSnapshot(`"export type Example = void;"`);
    });

    test("null", () => {
      expect(
        generateTypesSource({
          endpoints: {},
          types: {
            Example: NULL
          },
          description: {
            name: "name",
            description: "description"
          }
        })
      ).toMatchInlineSnapshot(`"export type Example = null;"`);
    });

    test("boolean", () => {
      expect(
        generateTypesSource({
          endpoints: {},
          types: {
            Example: BOOLEAN
          },
          description: {
            name: "name",
            description: "description"
          }
        })
      ).toMatchInlineSnapshot(`"export type Example = boolean;"`);
    });

    test("boolean constant", () => {
      expect(
        generateTypesSource({
          endpoints: {},
          types: {
            Example: booleanConstant(true)
          },
          description: {
            name: "name",
            description: "description"
          }
        })
      ).toMatchInlineSnapshot(`"export type Example = true;"`);
      expect(
        generateTypesSource({
          endpoints: {},
          types: {
            Example: booleanConstant(false)
          },
          description: {
            name: "name",
            description: "description"
          }
        })
      ).toMatchInlineSnapshot(`"export type Example = false;"`);
    });

    test("string", () => {
      expect(
        generateTypesSource({
          endpoints: {},
          types: {
            Example: STRING
          },
          description: {
            name: "name",
            description: "description"
          }
        })
      ).toMatchInlineSnapshot(`"export type Example = string;"`);
    });

    test("string constant", () => {
      expect(
        generateTypesSource({
          endpoints: {},
          types: {
            Example: stringConstant("some constant")
          },
          description: {
            name: "name",
            description: "description"
          }
        })
      ).toMatchInlineSnapshot(`"export type Example = \\"some constant\\";"`);
    });

    test("number", () => {
      expect(
        generateTypesSource({
          endpoints: {},
          types: {
            Example: NUMBER
          },
          description: {
            name: "name",
            description: "description"
          }
        })
      ).toMatchInlineSnapshot(`"export type Example = number;"`);
    });

    test("integer constant", () => {
      expect(
        generateTypesSource({
          endpoints: {},
          types: {
            Example: integerConstant(0)
          },
          description: {
            name: "name",
            description: "description"
          }
        })
      ).toMatchInlineSnapshot(`"export type Example = 0;"`);
      expect(
        generateTypesSource({
          endpoints: {},
          types: {
            Example: integerConstant(123)
          },
          description: {
            name: "name",
            description: "description"
          }
        })
      ).toMatchInlineSnapshot(`"export type Example = 123;"`);
      expect(
        generateTypesSource({
          endpoints: {},
          types: {
            Example: integerConstant(-1000)
          },
          description: {
            name: "name",
            description: "description"
          }
        })
      ).toMatchInlineSnapshot(`"export type Example = -1000;"`);
    });

    test("object", () => {
      expect(
        generateTypesSource({
          endpoints: {},
          types: {
            Example: objectType({})
          },
          description: {
            name: "name",
            description: "description"
          }
        })
      ).toMatchInlineSnapshot(`"export type Example = {};"`);
      expect(
        generateTypesSource({
          endpoints: {},
          types: {
            Example: objectType({
              singleField: NUMBER
            })
          },
          description: {
            name: "name",
            description: "description"
          }
        })
      ).toMatchInlineSnapshot(`
"export type Example = {
    singleField: number;
};"
`);
      expect(
        generateTypesSource({
          endpoints: {},
          types: {
            Example: objectType({
              field1: NUMBER,
              field2: STRING,
              field3: optionalType(BOOLEAN)
            })
          },
          description: {
            name: "name",
            description: "description"
          }
        })
      ).toMatchInlineSnapshot(`
"export type Example = {
    field1: number;
    field2: string;
    field3?: boolean;
};"
`);
    });

    test("array", () => {
      expect(
        generateTypesSource({
          endpoints: {},
          types: {
            Example: arrayType(STRING)
          },
          description: {
            name: "name",
            description: "description"
          }
        })
      ).toMatchInlineSnapshot(`"export type Example = string[];"`);
    });

    test("optional", () => {
      expect(
        generateTypesSource({
          endpoints: {},
          types: {
            Example: optionalType(STRING)
          },
          description: {
            name: "name",
            description: "description"
          }
        })
      ).toMatchInlineSnapshot(`"export type Example = string | void;"`);
    });

    test("union", () => {
      expect(
        generateTypesSource({
          endpoints: {},
          types: {
            Example: unionType(STRING, NUMBER, BOOLEAN)
          },
          description: {
            name: "name",
            description: "description"
          }
        })
      ).toMatchInlineSnapshot(
        `"export type Example = string | number | boolean;"`
      );
    });

    test("type reference", () => {
      expect(
        generateTypesSource({
          endpoints: {},
          types: {
            Example: typeReference("OtherType")
          },
          description: {
            name: "name",
            description: "description"
          }
        })
      ).toMatchInlineSnapshot(`"export type Example = OtherType;"`);
    });
  });
});
