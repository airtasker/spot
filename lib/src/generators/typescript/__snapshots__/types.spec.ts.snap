// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeScript types generator produces valid code multi-file 1`] = `
"export type CreateUserRequest = {
    name: string;
};

export type CreateUserResponse = {
    success: boolean;
};

export type GetUserResponse = {
    name: string;
};"
`;

exports[`TypeScript types generator produces valid code single-file 1`] = `
"export type CreateUserRequest = {
    name: string;
};

export type CreateUserResponse = {
    success: boolean;
};

export type GetUserResponse = {
    name: string;
};"
`;
